httpbp/middlewares.go

```diff
@@ -246,3 +255,103 @@ func SupportedMethods(method string, additional ...string) Middleware {
 		}
 	}
 }
+
+// statusCodeRecorder is used by RecordStatusCode to record the code passed to
+// a call to WriteHeader.
+type statusCodeRecorder struct {
+	http.ResponseWriter
+
+	code int
+}
+
+func (r *statusCodeRecorder) WriteHeader(code int) {
+	r.ResponseWriter.WriteHeader(code)
+	if r.code == 0 {
+		r.code = code
+	}
+}
+
+// getStatusCodeFamily takes an http status code and returns it as an "Nxx"
+// string. Returns "nan" if code > 999 and "-nan" if code < 100.
+func getStatusCodeFamily(code int) string {
+	family := code / 100
+	switch family {
+	case 1:
+		return "1xx"
+	case 2:
+		return "2xx"
+	case 3:
+		return "3xx"
+	case 4:
+		return "4xx"
+	case 5:
+		return "5xx"
+	case 6:
+		return "6xx"
+	case 7:
+		return "7xx"
+	case 8:
+		return "8xx"
+	case 9:
+		return "9xx"
+	default:
+		if family < 1 {
+			return "-nan"
+		}
+		return "nan"
+	}
+}
+
+// CounterGenerator is used by RecordStatusCode to create counters for recording
+// http response codes set by the server.
+type CounterGenerator interface {
+	Counter(name string) metrics.Counter
+}
```

Comment:
I really just added this to make it easier to test `RecordStatusCode`. I can remove it if we prefer to keep it cleaner and just use metricsbp.M or something different?

Reply:
I would rather either move this interface definition into `metricsbp` package, or just use `metricsbp.M`. I think I slightly prefer to just use `metricsbp.M`, as when we moved to prometheus it won't give us `metrics.Counter` anyways so this doesn't really give us any flexibility we wanted.

For testing you could have an unexported interface and an unexported version (`recordStatusCode(counters counterGenerator)`), and the exported version just do `return recordStatusCode(metricsbp.M)`.

Reply:
Good call